### PR TITLE
Winograd RxSf2x3 solver Update

### PR DESCRIPTION
**Pull request by shurale-nkn**
Winograd_v21 changes: added fp16 support.

WRW with stride 2 not supported.

FP16 convolution time improvements comparing with  fe4bb61

W | H | c | n | k | x | y | p | q | u | v | g |   | FWD | BWD | WRW
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
14 | 14 | 1024 | 32 | 1024 | 3 | 3 | 1 | 1 | 2 | 2 | 32 |   | 0% | -34% | 0%
7 | 7 | 1024 | 32 | 1024 | 3 | 3 | 1 | 1 | 1 | 1 | 32 |   | -40% | -75% | -14%
56 | 56 | 256 | 32 | 256 | 3 | 3 | 1 | 1 | 2 | 2 | 32 |   | -58% | 0% | 0%
14 | 14 | 512 | 32 | 512 | 3 | 3 | 1 | 1 | 1 | 1 | 32 |   | 0% | -25% | 0%
28 | 28 | 512 | 32 | 512 | 3 | 3 | 1 | 1 | 2 | 2 | 32 |   | 0% | -44% | 0%
14 | 14 | 1024 | 32 | 1024 | 3 | 3 | 1 | 1 | 1 | 1 | 64 |   | 0% | -20% | 0%
28 | 28 | 1024 | 32 | 1024 | 3 | 3 | 1 | 1 | 2 | 2 | 64 |   | 0% | -52% | 0%
14 | 14 | 2048 | 32 | 2048 | 3 | 3 | 1 | 1 | 2 | 2 | 64 |   | 0% | -39% | 0%
7 | 7 | 2048 | 32 | 2048 | 3 | 3 | 1 | 1 | 1 | 1 | 64 |   | -53% | -80% | -31%
56 | 56 | 512 | 32 | 512 | 3 | 3 | 1 | 1 | 2 | 2 | 64 |   | -46% | 0% | 0%
14 | 14 | 1024 | 32 | 1024 | 3 | 3 | 1 | 1 | 1 | 1 | 32 |   | -68% | -67% | 0%
28 | 28 | 1024 | 32 | 1024 | 3 | 3 | 1 | 1 | 2 | 2 | 32 |   | -75% | -63% | 0%
14 | 14 | 2048 | 32 | 2048 | 3 | 3 | 1 | 1 | 2 | 2 | 32 |   | 0% | -50% | 0%
7 | 7 | 2048 | 32 | 2048 | 3 | 3 | 1 | 1 | 1 | 1 | 32 |   | -67% | -95% | -22%
28 | 28 | 512 | 32 | 512 | 3 | 3 | 1 | 1 | 1 | 1 | 32 |   | 0% | -32% | 0%
56 | 56 | 512 | 32 | 512 | 3 | 3 | 1 | 1 | 2 | 2 | 32 |   | 0% | -41% | 0%
13 | 13 | 192 | 512 | 384 | 3 | 3 | 1 | 1 | 1 | 1 | 1 |   | -4% | 0% | -38%
13 | 13 | 384 | 512 | 256 | 3 | 3 | 1 | 1 | 1 | 1 | 1 |   | -4% | -40% | -33%
14 | 14 | 512 | 128 | 512 | 3 | 3 | 1 | 1 | 1 | 1 | 1 |   | 0% | -4% | 0%
28 | 28 | 512 | 128 | 512 | 3 | 3 | 1 | 1 | 1 | 1 | 1 |   | -5% | 0% | 0%
17 | 17 | 192 | 64 | 192 | 7 | 1 | 0 | 3 | 1 | 1 | 1 |   | 0% | 0% | -67%
17 | 17 | 192 | 64 | 224 | 7 | 1 | 0 | 3 | 1 | 1 | 1 |   | -1% | -1% | -58%
17 | 17 | 224 | 64 | 224 | 7 | 1 | 0 | 3 | 1 | 1 | 1 |   | 0% | -1% | -67%
17 | 17 | 224 | 64 | 256 | 1 | 7 | 3 | 0 | 1 | 1 | 1 |   | 0% | -1% | -42%

### CODE REVIEW

**Comment by daniellowell** [comment]
Formatting required.

File: src/kernels/Conv_Winograd_v21_1_0_metadata.inc [hunk 1]
```diff
@@ -117,15 +98,60 @@ amdhsa.kernels:
...
 
 total_sgpr_count = .amdgcn.next_free_sgpr + 4 // vcc, xnack
 
-METADATA_WRAPPER total_sgpr_count,.amdgcn.next_free_vgpr,workgroup_size_x
+METADATA_WRAPPER total_sgpr_count,.amdgcn.next_free_vgpr,workgroup_size_x, <\kernel_name>
+
+.endm
+
+.macro KERNEL_PROLOG kernel_name_postfix
+	PROLOG_KERNEL_DESCRIPTOR miopenSp3AsmConv_v21_1_0_\kernel_name_postfix
```
[the remark this answers is not in the document]
Comment (shurale-nkn) [reply src/kernels/Conv_Winograd_v21_1_0_metadata.inc:151]: No, it shouldn't. This metadata for version 21_1_0, i have no guarantees that nothing will change in the next one.

[the remark this answers is not in the document]
Comment (Kirpich30000) [reply src/kernels/Conv_Winograd_v21_1_0_metadata.inc:151]: It's a bad design pattern that a string is assembled from the pieces from different files, and it doesn't resolve any dependency problems. It makes code less obvious - you can't determine the name looking only on a single file and have to remember what is going on in another place. The macro on the other hand became less handy - even if other kernels will support the same ABI you will not be able to use that macro. If you want to explicitly emphasize ABI version dependency - use v21 in a macro name. That way:
- code became clearer:
	- exactly the same string that you pass as an argument would be used, no hidden modifications
	- unnecessary macros in Conv_Winograd_v21_1_0_metadata.inc that do nothing but concatenate strings could be removed
	- ABI version became obvious from the macro name
- macro became reusable. If the following Conv_Winograd_*.s will support the same argument block that would be clearly stated in the code.

**Comment by daniellowell** [comment]
Runtime issues:
```
[100%] Built target test_conv2d

Scanning dependencies of target test_conv_for_implicit_gemm

../bin/test_conv2d --half --cmode conv --pmode default --group-count 1 --input 64, 16, 28, 28 --weights 192, 16, 3, 3 --pads_strides_dilations 0 0 2 2 1 1 --trans_output_pads 0 0 

MIOpen(HIP): Warning [KernDb] Invalid fields in table: kern_db disabling access to 

terminate called after throwing an instance of 'miopen::Exception'

  what():  /var/jenkins/workspace/MLLibs_MIOpen_winograd_21/build/kernel.cpp:176: Failed to load kernel source: CONV_WINOGRAD_V21_1_0_GFX9_FP16_FP16_PK_STRIDE2

Aborted (core dumped)
```

http://micimaster.amd.com/blue/organizations/jenkins/MLLibs%2FMIOpen/detail/winograd_21/6/pipeline/

Also, please merge the latest develop into this branch.

File: src/solver/conv_winoRxS_f2x3.cpp [hunk 2]
```diff
@@ -196,6 +196,11 @@ void PerformanceConfigConvBinWinogradRxSf2x3::EuristicInit(const ConvolutionCont
 {
     const auto n_inputs_per_group  = config.n_inputs / config.group_counts,
                n_outputs_per_group = config.n_outputs / config.group_counts;
+    if(config.group_counts == 1)
```
Comment (atamazov) [inline src/solver/conv_winoRxS_f2x3.cpp:199]: Bugfix?

File: src/solver/conv_winoRxS_f2x3.cpp [hunk 3]
```diff
@@ -480,7 +485,7 @@ bool ConvBinWinogradRxSf2x3::IsApplicable(const ConvolutionContext& params) cons
 {
     if(!params.Is2d())
         return false;
-    if(!params.IsFp32())
+    if(!(params.IsFp32() || params.IsFp16()))
```
Comment (atamazov) [inline src/solver/conv_winoRxS_f2x3.cpp:488]: Same coverage as FP32, none new limitations like `(c % 2 == 0)`?

**Comment by atamazov** [comment]
Please fix missing EOLs in assembly files (github is complaining).

**Comment by atamazov** [comment]
> WRW with stride 2 not supported.

This sound like a new limitation, or like fp32 WrW supports stride 2 but fp16 doesn't, but this is not so right?

**Comment by atamazov** [comment]
What is urgency/importance of this?

File: src/solver/conv_winoRxS_f2x3.cpp (hunk 2 above)
Comment (shurale-nkn) [reply src/solver/conv_winoRxS_f2x3.cpp:199]: performance fix 😅.
EuristicInit can't calculate fastest configuration in some cases. But this is fastest configuration  for all group_counts == 1.

**Comment by daniellowell** [comment]
@JehandadKhan @alexandraBara @cderb Added you guys for visibility.
@shurale-nkn I will merge this PR as soon as I cut the new release branch.

File: src/solver/conv_winoRxS_f2x3.cpp (hunk 3 above)
Comment (shurale-nkn) [reply src/solver/conv_winoRxS_f2x3.cpp:488]: Yes, same as for FP32.

**Comment by shurale-nkn** [comment]
> > WRW with stride 2 not supported.
> 
> This sounds like a new limitation, or like fp32 WrW supports stride 2 but fp16 doesn't, but this is not so right?

Filter stride==2 is not supported for both data types.

File: src/solver/conv_winoRxS_f2x3.cpp [hunk 4]
```diff
@@ -602,22 +607,40 @@ ConvBinWinogradRxSf2x3::GetSolution(const ConvolutionContext& params,
...
+    if(params.IsFp32())
+        kernel_postfix += "_fp32";
+    else
+    {
+        kernel_postfix += "_fp16";
+        const auto name = params.GetStream().GetDeviceName();
+        if(StartsWith(name, "gfx906") || StartsWith(name, "gfx908"))
+            kernel_postfix += "_dot2_edc";
+        else
+            kernel_postfix += "_pk";
```
Comment (atamazov) [inline src/solver/conv_winoRxS_f2x3.cpp:623]: This is pure FP16 kernel. We are not using pure FP16 kernels. Please disable and remove.

**Pull request merged**